src/apps/polymer/IncentivizedPolymerEscrow.sol

```diff
@@ -0,0 +1,175 @@
+    error NotImplemented();
+
+    // Internal function for gas savings.
+    function _UNIQUE_SOURCE_IDENTIFIER() internal view returns (bytes32) {
+        return bytes32(block.chainid);
+    }
+
+    // Expose internal function.
+    function UNIQUE_SOURCE_IDENTIFIER() external view returns (bytes32) {
+        return _UNIQUE_SOURCE_IDENTIFIER();
+    }
+
+    struct VerifiedMessageHashContext {
+        bytes32 chainIdentifier;
+        bytes implementationIdentifier;
+    }
+
+    mapping(bytes32 => VerifiedMessageHashContext) public isVerifiedMessageHash;
+
+    constructor(address sendLostGasTo, address messagingProtocol)
+        IMETimeoutExtension(sendLostGasTo)
+        IbcMwUser(messagingProtocol)
+    {}
+
+    function estimateAdditionalCost() external pure returns (address asset, uint256 amount) {
+        asset = address(0);
+        amount = 0;
+    }
+
+    function _getMessageIdentifier(bytes32 destinationIdentifier, bytes calldata message)
+        internal
+        view
+        override
+        returns (bytes32)
+    {
+        return keccak256(
+            abi.encodePacked(bytes32(block.number), _UNIQUE_SOURCE_IDENTIFIER(), destinationIdentifier, message)
+        );
+    }
+
+    /// @notice This function is used to allow acks to be executed twice (if the first one ran out of gas)
+    /// This is not intended to allow processPacket to work.
+    function _verifyPacket(bytes calldata, /* _metadata */ bytes calldata _message)
+        internal
+        view
+        override
+        returns (bytes32 sourceIdentifier, bytes memory implementationIdentifier, bytes calldata message_)
+    {
+        sourceIdentifier = isVerifiedMessageHash[keccak256(_message)].chainIdentifier;
+        implementationIdentifier = isVerifiedMessageHash[keccak256(_message)].implementationIdentifier;
+
+        if (sourceIdentifier == bytes32(0)) revert NonVerifiableMessage();
+
+        message_ = _message;
+    }
+
+    /// @dev Disable processPacket
+    function processPacket(
+        bytes calldata, /* messagingProtocolContext */
+        bytes calldata, /* rawMessage */
+        bytes32 /* feeRecipitent */
+    ) external payable override {
+        revert NotImplemented();
+    }
+
+    // packet.srcPortAddr is the IncentivizedPolymerEscrow address on the source chain.
+    // packet.destPortAddr is the address of this contract.
+    // channelId: the universal channel id on the source chain, which can be used to identify the source chain.
+    function onRecvUniversalPacket(bytes32 channelId, UniversalPacket calldata packet)
+        external
+        onlyIbcMw
+        returns (AckPacket memory)
+    {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes memory sourceImplementationIdentifier = abi.encodePacked(packet.srcPortAddr);
+
+        bytes memory receiveAck = _handleMessage(
+            bytes32(abi.encodePacked(packet.srcPortAddr)),
```

Comment:
This line should be chainIdentifier similarly to how chainIdentifier is derived on ack: `bytes32 chainIdentifier = _getChannelId(channelId);`

Reply:
Following our discussion, treat channelId as chainIdentifier